### ISSUE TITLE
`needless_update` can be good with `..T::default()` for compatibility

### ISSUE THREAD

**Issue by wchargin** [rust-lang/rust-clippy#6323]
Clippy’s `needless_update` lint [fires on the following code][play]:

```rust
#[derive(Default)]
struct S {
    x: i32,
}
fn foo() -> S {
    S {
        x: 123,
        ..S::default()
    }
}
```

```
warning: struct update has no effect, all the fields in the struct have already been specified
  --> src/lib.rs:10:11
   |
10 |         ..S::default()
   |           ^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::needless_update)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_update

warning: 1 warning emitted
```

This is reasonable and sensible. But there are cases where the above
code is what I want. Specifically: if I expect to add more fields to `S`
later with default values, then including `S::default()` means that I
won’t have to update this call site.

I understand that that kind of thing is not entirely Rustic: having the
compiler show you all the call sites to update is powerful. One case
where this is nevertheless quite valuable is when `S` is a generated
binding for a [protocol buffer][pb] message type. Ease of evolution is a
core design principle of protocol buffers; in protobuf, all types have
default values, so adding a field is a backward-compatible change. In
particular, it’s important in a large codebase to be able to add a field
to a protobuf and not have to go through hundreds of call sites to
update code.

In fact, I would actually _like_ a lint that fired if you construct a
struct at type `S: prost::Message` _without_ including `..S::default()`,
to _ensure_ that such proto updates will proceed smoothly.

From another perspective, I view this as a kind of `#[non_exhaustive]`
that lets you construct a value in exchange for accepting that future
changes may introduce new fields.

Suggestions:

  - Could we make `needless_update` not fire when the base struct is a
    call to `<T as Default>::default`? That way, I can keep the lint on
    and get useful info about non-protobuf needless updates.
  - If that’s not desirable, could we add a note to the “Known problems”
    page for `needless_updates`?

[play]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fc2ab5b8609e5053e0c6a57fddf19e8f
[pb]: https://developers.google.com/protocol-buffers/

**Comment by flip1995** [comment]
IMO, the lint behavior is correct to lint this. You usually want rustc to tell you if something is added to a struct. 

I think the best thing we can do here would be to not trigger on structs that are marked as `non_exhaustive`.

**Comment by heaths** [comment]
> I think the best thing we can do here would be to not trigger on structs that are marked as `non_exhaustive`.

See rust-lang/lang-team#143. There's been some discussion, but not traction.

Anyone writing, say, crates for cloud services would benefit as models tend to change (hopefully only additions) so for any crates using those libs, being able to specify `..Default::default()` is helpful to avoid breaking changes. If you really needed to set a field, chances are you either requested the feature on the service or have been following it. Expecting clippy to tell you of some arbitrary field being added is probably not interesting unless you actually need it, especially if it's optional/can be defaulted.
